fix(core): use short timeouts for periodic `KVS.setRecord` calls

B4nan opened the pull request:
Sometimes the `setRecord` calls might get stuck and since we issue a new one every 60s (via `persistState` event), and the default API call timeout is 6 minutes, we might end up overloading the API.

This PR disables retries on those calls in case of a timeout, and overrides the default 6-minute timeout to half of the persist state interval (so to 30s by default).

Related: https://apify.slack.com/archives/C0L33UM7Z/p1747043120071539

B4nan commented:
e2e tests running at https://github.com/apify/crawlee/actions/runs/14994236197

B4nan commented:
e2e tests passed

barjin commented on packages/core/src/storages/key_value_store.ts:
```diff
@@ -273,9 +273,16 @@ export class KeyValueStore {
 
+        // use half the interval of `persistState` to avoid race conditions
+        const persistStateIntervalMillis = this.config.get('persistStateIntervalMillis')!;
+        const timeoutSecs = persistStateIntervalMillis / 2_000;
+
         this.config.getEventManager().on('persistState', async () => {
             for (const [key, value] of this.cache) {
-                await this.setValue(key, value);
+                await this.setValue(key, value, {
+                    timeoutSecs,
+                    doNotRetryTimeouts: true,
+                });
```
Do I understand correctly that the issue was too many concurrent writes _to the same key_, right? 

![image](https://github.com/user-attachments/assets/cb96a1a3-e078-4435-89b3-e37dec372646)

So while this code might still intermingle multiple "epochs" of writes together (see "timeouting" red key-value `cache` pairs from the old `persistState` event and blue writes from the new `persistState` event), it will never issue more than one concurrent write to the same key (because of the `/ 2` timeout)?

B4nan replied:
I see, normally there will be just one key in that cache object (the one for `useState`), so I guess we should be safe. Maybe I should rewrite this to use `Promise.all` and fire all those concurrently, since every entry will have a different key.

barjin reviewed (approved):
Looks alright to me :+1: a small nit / question regarding the loop in `ensurePersistStateEvent`.

barjin commented on packages/core/src/session_pool/session_pool.ts:
```diff
@@ -367,7 +367,14 @@ export class SessionPool extends EventEmitter {
+        const timeoutSecs = persistStateIntervalMillis / 2_000;
+        await this.keyValueStore.setValue(this.persistStateKey, this.getState(), {
+            timeoutSecs,
+            doNotRetryTimeouts: true,
```
nit: how about `retryOnTimeouts:` or `retryTimeouts: false`?

B4nan replied:
this is an existing option in the client, no plans to introduce a new name for it here, IMO that would make things only more confusing

(i don't like its name too, but still, lets stay consistent)

B4nan commented:
rewritten to use `Promise.all` and also added `.catch` to all those 3 places, since i believe a rejected promise was the cause for the actor runs to fail completely 

e2e tests running again at https://github.com/apify/crawlee/actions/runs/14999353250

barjin commented:
> httpbin outage

How come we don't use `httpbin.apify.actor` here :face_with_spiral_eyes: ? The OG httpbin has indeed been quite flaky recently. I added it to [this issue](https://github.com/apify/crawlee/issues/2906#issuecomment-2878774869) and will change this today (should be a fairly easy fix).

tobice reviewed (approved):
Looks good :+1: 

I'm really curious to see if this helps.

B4nan commented:
I am rather sure it will help. Not so sure if it will resolve it completely, but there are 3 separate improvements, all of them will make things better
- `useState` is persisted in parallel instead of one by one
- all periodic state persistence from those 3 places is catching potential errors (so no failed actor runs because of those)
- all periodic state persistence skip retries on timeouts

I'll ship this to stable right ahead.
